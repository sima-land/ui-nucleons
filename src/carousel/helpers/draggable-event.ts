import { IPoint } from '../../helpers/point';

interface Data {
  offset: IPoint;
  client: IPoint;
}

/**
 * Нестандартное событие перетаскивания.
 */
export class DraggableEvent implements Data {
  prevented: boolean;
  offset: IPoint;
  client: IPoint;

  /**
   * Конструктор.
   * @param props Свойства.
   * @param props.offset Смещение контента.
   * @param props.client Клиентская позиция.
   */
  constructor(props: Data) {
    this.prevented = false;
    this.offset = props.offset;
    this.client = props.client;
  }

  /**
   * Помечает событие как отмененное.
   */
  preventDefault() {
    this.prevented = true;
  }
}
