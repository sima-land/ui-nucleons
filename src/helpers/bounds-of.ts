/**
 * Возвращает ограничивающий прямоугольник элемента.
 * @param element Элемент.
 * @return Ограничивающий прямоугольник.
 */
export function boundsOf<T>(element: T): T extends Element ? DOMRect : null {
  return (element instanceof Element ? element.getBoundingClientRect() : null) as T extends Element
    ? DOMRect
    : null;
}
