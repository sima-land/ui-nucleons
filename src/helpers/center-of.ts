import { Point } from './point';

/**
 * Возвращает координаты центра прямоугольника.
 * @param rect Прямоугольник.
 * @return Центр.
 */
export function centerOf(rect: DOMRect) {
  return Point((rect.right + rect.left) / 2, (rect.bottom + rect.top) / 2);
}
