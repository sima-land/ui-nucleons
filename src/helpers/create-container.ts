export interface Container<T> {
  get: () => T | undefined;
  set: (v: T) => void;
  setObserver: (fn: Observer<T> | null) => void;
}

type Observer<T> = (prev: T | undefined, next: T | undefined) => void;

/**
 * Фабрика контейнеров для произвольных значений.
 * @param initialValue Начальное значение в контейнере.
 * @return Контейнер.
 */
export function createContainer<T = any>(initialValue?: T): Container<T> {
  let value: T | undefined = initialValue;
  let observe: Observer<T> | null;

  /**
   * Геттер значения.
   * @return Значение.
   */
  const get = (): T | undefined => value;

  /**
   * Сеттер значения.
   * @param newValue Новое значение.
   */
  const set = (newValue: T) => {
    const oldValue: T | undefined = value;
    value = newValue;

    !Object.is(newValue, oldValue) && observe?.(newValue, oldValue);
  };

  /**
   * Сеттер функции-наблюдателя за изменением значения.
   * @param newObserve Функция-наблюдатель.
   */
  const setObserver = (newObserve: Observer<T> | null) => {
    observe = newObserve;
  };

  return { get, set, setObserver };
}
