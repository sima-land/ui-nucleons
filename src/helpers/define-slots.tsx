import {
  Children,
  ComponentProps,
  ElementType,
  ReactElement,
  ReactNode,
  isValidElement,
} from 'react';

export type SlotsDict = Record<string, ElementType>;

export type Slots<M extends SlotsDict> = {
  [key in keyof M]: ReactElement<ComponentProps<M[key]>, M[key]> | undefined;
};

/**
 * Определяет, является ли React-элемент элементом заданного типа.
 * @param value Элемент.
 * @param type Тип.
 * @return Результат.
 */
function isComponentOfType<T extends ElementType>(
  value: ReactElement,
  type: T,
): value is ReactElement<any, T> {
  return value.type === type;
}

/**
 * Распределяет дочерние элементы по слотам.
 * @param children Дочерние элементы.
 * @param dict Словарь типов элементов.
 * @return Дочерние элементы разбитые по слотам.
 */
export function defineSlots<T extends SlotsDict>(children: ReactNode, dict: T): Slots<T> {
  const allKeys: Array<keyof T> = Object.keys(dict);

  const slots: Slots<T> = Children.toArray(children).reduce<Slots<T>>((result, item) => {
    if (isValidElement(item)) {
      for (const key of allKeys) {
        if (!result[key] && isComponentOfType(item, dict[key])) {
          result[key] = item;
          break;
        }
      }
    }

    return result;
  }, {} as any);

  return slots;
}
