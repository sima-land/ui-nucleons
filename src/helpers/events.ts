import type { MouseEvent as ReactMouseEvent, TouchEvent as ReactTouchEvent } from 'react';
import { Point } from './point';

export type EventWithPosition<T extends HTMLElement = HTMLElement> =
  | MouseEvent
  | TouchEvent
  | ReactMouseEvent<T>
  | ReactTouchEvent<T>;

/**
 * Проверяет, является ли событие событием нажатия основной кнопки мыши.
 * @param event Событие.
 * @return Является ли.
 */
export function isMainMouseButton(event: any) {
  return event?.button === 0;
}

/**
 * Проверяет, является ли переданное событие touch-событием.
 * @param event Событие.
 * @return Является ли переданное событие touch-событием.
 */
export function isTouchEvent(event: any): event is TouchEvent | ReactTouchEvent {
  return Boolean(event?.touches);
}

/**
 * Берет данные позиции из события.
 * @param event Touch- или Mouse-событие.
 * @return Данные позиции.
 */
export function getEventClientPos<T extends HTMLElement = HTMLElement>(
  event: EventWithPosition<T>,
) {
  const source = isTouchEvent(event) ? event.touches[0] : event;

  return Point(source?.clientX, source?.clientY);
}

/**
 * Вызывает событие input на элементе input/textarea/select.
 * Полезен когда необходимо сообщить React'у что был произведен ввод значения пользователем.
 * За основу взята реализация отсюда: https://stackoverflow.com/a/47409362/13166471.
 * @param element Элемент.
 * @param value Значение.
 */
export function triggerInput(
  element: HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement,
  value: string,
): void {
  const constructors = [HTMLInputElement, HTMLSelectElement, HTMLTextAreaElement] as const;

  let constructor:
    | typeof HTMLInputElement
    | typeof HTMLTextAreaElement
    | typeof HTMLSelectElement
    | null = null;

  for (const item of constructors) {
    if (element instanceof item) {
      constructor = item;
    }
  }

  if (constructor) {
    const setValue = Object.getOwnPropertyDescriptor(constructor.prototype, 'value')?.set;

    if (setValue) {
      setValue.call(element, value);
      element.dispatchEvent(new Event('input', { bubbles: true }));
    }
  }
}
