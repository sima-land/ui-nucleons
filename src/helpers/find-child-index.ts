/**
 * Выполняет цикл для поиска родственного элемента по заданным условиям.
 * @param props Свойства.
 * @param props.target Элемент для которого нужно найти родственный элемент.
 * @param props.startIndex Индекс элемента с которого нужно начать поиск.
 * @param props.increment Значение инкремента.
 * @param props.defaultResult Результат в случае если ни один элемент не прошел проверку.
 * @param props.isSuitable Должна определить удовлетворяет ли элемент условию.
 * @param props.needBreakLoop Должна определить нужно ли останавливать цикл.
 * @return Индекс дочернего элемента.
 */
export function findChildIndex({
  target,
  startIndex = 0,
  increment = 1,
  defaultResult = -1,
  isSuitable,
  needBreakLoop = a => a,
}: {
  target?: HTMLElement | null | undefined;
  startIndex?: number;
  increment?: number;
  defaultResult?: number;
  isSuitable?: (el: Element) => boolean;
  needBreakLoop?: (passed: boolean) => boolean;
}): number {
  let result = defaultResult;

  if (target && target.children && isSuitable !== undefined) {
    const { children } = target;

    for (let i = startIndex; i >= 0 && i < children.length; i += increment) {
      const sibling = children[i];
      const passed = isSuitable(sibling);

      if (passed) {
        result = i;
      }
      if (needBreakLoop(passed)) {
        break;
      }
    }
  }

  return result;
}
