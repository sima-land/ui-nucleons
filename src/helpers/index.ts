// ВАЖНО: только определенные утилиты являются публичными, они экспортируются здесь
export { defineSlots } from './define-slots';
export { findOffsetParent } from './find-offset-parent';
export { formatDate, createDateFormatter } from './format-date';
export { getDateIntervalData } from './get-date-interval-data';
export { getDeclination } from './get-declination';
export { getPositionedParentOffset } from './get-positioned-parent-offset';
export { getScrollParent } from './get-scroll-parent';
export { isBrowser } from './is-browser';
export { isFullyScrolled } from './is-fully-scrolled';
export { isTouchDevice } from './is-touch-device';
export { on } from './on';
export { scrollToChild } from './scroll-to-child';
export { useLayer, LayerProvider } from './layer';
export { mergeRefs } from './merge-refs';
