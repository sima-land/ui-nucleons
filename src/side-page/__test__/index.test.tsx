import { render, fireEvent } from '@testing-library/react';
import { SidePage } from '..';
import { LayerProvider, useLayer } from '../../helpers/layer';

describe('SidePage', () => {
  it('should renders correctly', async function () {
    const { container, rerender } = render(
      <SidePage withTransitions shown={false}>
        <SidePage.Header title='Test title' />
        <SidePage.Body>
          <h1>Test body</h1>
        </SidePage.Body>
        <SidePage.Footer>
          <p>Test footer</p>
        </SidePage.Footer>
      </SidePage>,
    );

    expect(container).toMatchSnapshot();

    rerender(
      <SidePage withTransitions shown>
        <SidePage.Header title='Test title' />
        <SidePage.Body>
          <h1>Test body</h1>
        </SidePage.Body>
        <SidePage.Footer>
          <p>Test footer</p>
        </SidePage.Footer>
      </SidePage>,
    );

    // ждем окончания анимации
    await new Promise(done => setTimeout(done, 1000));

    expect(container).toMatchSnapshot();
    expect(container.querySelectorAll('.enter-active')).toHaveLength(0);
  });

  it('should handle header props', async function () {
    const onBack = jest.fn();
    const onClose = jest.fn();

    const { getByTestId } = render(
      <SidePage shown>
        <SidePage.Header title='Test title' onBack={onBack} onClose={onClose} />
        <SidePage.Body>
          <h1>Test body</h1>
        </SidePage.Body>
        <SidePage.Footer>
          <p>Test footer</p>
        </SidePage.Footer>
      </SidePage>,
    );

    expect(onBack).toHaveBeenCalledTimes(0);
    fireEvent.click(getByTestId('top-bar:back'));
    expect(onBack).toHaveBeenCalledTimes(1);

    expect(onClose).toHaveBeenCalledTimes(0);
    fireEvent.click(getByTestId('top-bar:close'));
    expect(onClose).toHaveBeenCalledTimes(1);
  });

  it('should increment layer by 100', () => {
    function TestComponent() {
      const layer = useLayer();
      return <div data-testid='test-component' data-layer={layer} />;
    }

    const { getByTestId } = render(
      <LayerProvider value={20}>
        <SidePage shown>
          <SidePage.Body>
            <TestComponent />
          </SidePage.Body>
        </SidePage>
      </LayerProvider>,
    );

    expect(getByTestId('test-component').getAttribute('data-layer')).toBe('120');
  });
});
