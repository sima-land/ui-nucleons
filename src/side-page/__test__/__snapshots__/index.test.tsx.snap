// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidePage should renders correctly 1`] = `<div />`;

exports[`SidePage should renders correctly 2`] = `
<div>
  <div
    class="root"
    data-testid="modal-overlay"
    style="--transition-duration: 300ms; z-index: 100; --vh: 7.68px;"
  >
    <div
      class="side-page size-m"
      data-testid="side-page"
    >
      <div
        class="root size-unset header no-subtitle"
        data-testid="top-bar"
      >
        <div
          class="main"
        >
          <div
            class="title"
          >
            Test title
          </div>
        </div>
      </div>
      <div
        class="body"
      >
        <h1>
          Test body
        </h1>
      </div>
      <div
        class="root rounds-md__b footer"
        data-testid="bottom-bar"
      >
        <p>
          Test footer
        </p>
      </div>
    </div>
  </div>
</div>
`;
